plutus-contract/src/Plutus/Contract/Marconi/Handler.hs

```diff
@@ -0,0 +1,176 @@
+import Plutus.ChainIndex.Types (ChainSyncBlock (..), Point (Point, PointAtGenesis), Tip (TipAtGenesis), citxCardanoTx,
+                                tipAsPoint)
+import Plutus.V2.Ledger.Api (TxOutRef (TxOutRef))
+
+{- Developer note:
+   How to add new indexer to support new effect?
+       1. add the indexer MVar to `ChainIndexIndexersMvar` and the corresponding indexer to `ChainIndexIndexers`
+       2. edit `getChainIndexIndexers` and `putChainIndexIndexers` accordingly
+       3. generate `MarconiEffect` on the appropriate queries of `ChainIndexQueries` in `handleQuery`
+       4. Add the indexer update in the control operations
+       5. Add the `handleMarconiQuery` for the new effect in `handleChainIndexEffects`
+-}
+
+data ChainIndexIndexers -- We don't use `newtype` since other indexers will be needed
+   = ChainIndexIndexers
+       { _utxosIndexer :: State UtxoHandle
+       }
+
+makeLenses ''ChainIndexIndexers
+
+data ChainIndexIndexersMVar -- We don't use `newtype` since other indexers will be needed
+   = ChainIndexIndexersMVar
+       { _utxosIndexerMVar :: MVar (State UtxoHandle)
+       }
+
+makeLenses ''ChainIndexIndexersMVar
+
+getChainIndexIndexers :: ChainIndexIndexersMVar -> IO ChainIndexIndexers
+getChainIndexIndexers mvarCi =
+    ChainIndexIndexers <$> readMVar (mvarCi ^. utxosIndexerMVar)
+
+putChainIndexIndexers :: ChainIndexIndexers -> ChainIndexIndexersMVar -> IO ()
+putChainIndexIndexers ci mvarCi = do
+    putMVar (mvarCi ^. utxosIndexerMVar) (ci ^. utxosIndexer)
+
+
+
+data MarconiEffect handle r where
+  QueryIndexer :: StorableQuery handle -> MarconiEffect handle (StorableResult handle)
+
+makeEffect ''MarconiEffect
+
+handleMarconiQuery ::
+    ( LastMember IO effs
+    , Member (Eff.State ChainIndexIndexers) effs
+    , StorableMonad handle ~ IO
+    , HasPoint (StorableEvent handle) (StorablePoint handle)
+    , Ord (StorablePoint handle)
+    , Queryable handle
+    )
+    => Lens' ChainIndexIndexers (State handle) -> MarconiEffect handle ~> Eff effs
+handleMarconiQuery l (QueryIndexer q) = do
+  ci <- Eff.get
+  liftIO $ query QEverything (ci ^. l) q
+
+getUtxoSetAtAddress
+  :: forall effs.
+    ( Member (MarconiEffect UtxoHandle) effs
+    )
+  => PageQuery TxOutRef
+  -> CardanoAddress
+  -> Eff effs UtxosResponse
+getUtxoSetAtAddress pageQuery addrInEra = let
+    toTxOutRef utxo = TxOutRef
+                          (fromCardanoTxId $ utxo ^. txId)
+                          (toInteger . (\(TxIx x) -> x) $ utxo ^. txIx)
+    addr = case addrInEra of
+                AddressInEra ByronAddressInAnyEra addr'    -> toAddressAny addr'
+                AddressInEra (ShelleyAddressInEra _) addr' -> toAddressAny addr'
+    in UtxosResponse TipAtGenesis
+           . pageOf pageQuery
+           . Set.fromList
+           . fmap (views urUtxo toTxOutRef)
+           . getUtxoResult
+           <$> queryIndexer (UtxoAddress addr)
+
+
+getUtxoEvents
```

Comment:
There is already a `getUtxoEvents` in the Utxo indexer module.

Reply:
Yes, but we need one for `CardanoTx` and we can't directly unwrap `[CardanoTx]` in a `[Tx era]` as we can't ensure that all the `CardanoTx` contain `Tx` from the same `era`.

Reply:
Noooo :(